### PR TITLE
Noahcaldwell/batting stats view

### PR DESCRIPTION
**Pull request by Debob123**
# Overview
added script that creates batting_stats_view for database with mostly accurate data

## Changes
added view file in dbsetup directory that includes init.py file that initializes view and script that creates view 

## Proof of Life
Loaded query into SQL to prove that it creates view correctly and selected specific values from houston astros in 2022 to see if it pulled correct data.

Icko:
pol-bsv

### CODE REVIEW

File: dbSetup/models/views/batting_stats_view.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        b.b_SB AS SB,
+        (b.b_BB / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) * 100 AS `BB%`,
+        (b.b_SO / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) * 100 AS `K%`,
+        (b.b_H - b.b_HR) / (b.b_AB - b.b_SO - b.b_HR + b.b_SF) AS BABIP,
+        (b.b_H / b.b_AB) AS AVG,
+        ((b.b_H - (b.b_2B + b.b_3B + b.b_HR)) + (2 * b.b_2B) + (3 * b.b_3B) + (4 * b.b_HR)) / b.b_AB AS SLG,
+        (((b.b_H - (b.b_2B + b.b_3B + b.b_HR)) + (2 * b.b_2B) + (3 * b.b_3B) + (4 * b.b_HR)) / b.b_AB) - (b.b_H / b.b_AB) AS ISO,
+        (b.b_H - (b.b_2B + b.b_3B + b.b_HR)) AS b_1B
+    FROM batting b
+    JOIN player p ON b.playerID = p.playerID
```
Comment (cateburglar) [inline dbSetup/models/views/batting_stats_view.py:61]: it's "JOIN people p..." not "player p"

File: dbSetup/models/views/batting_stats_view.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    JOIN appearances a ON b.playerID = a.playerID
+) subquery;
+"""
+
+# Execute the raw SQL
+session.commit()
```
Comment (cateburglar) [inline dbSetup/models/views/batting_stats_view.py:67]: this does not execute the sql. you need to do session.execute(----) first. as your code stands right now, nothing changes in the database

File: dbSetup/models/views/batting_stats_view.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+# Generate engine string from utility function
+database_values = {
+    "dialect": "sqlite",  # Replace with your actual database dialect
+    "database": "your_database.db",  # Replace with your actual database name
+}
+
+
+# Define the SQL query
+create_view_query = """
```
Comment (cateburglar) [inline dbSetup/models/views/batting_stats_view.py:14]: this query is not functional-- i'm not going to go through and debug it all myself so could you test this using the database on your terminal and then make the necessary changes

File: dbSetup/models/views/batting_stats_view.py (hunk 2 above)
Comment (ickoxii) [reply dbSetup/models/views/batting_stats_view.py:67]: update this query with whatever was in that sql file you loaded earlier. that seemed to calculate stats

File: dbSetup/models/views/batting_stats_view.py (hunk 3 above)
Comment (ickoxii) [reply dbSetup/models/views/batting_stats_view.py:14]: update this query with whatever was in that sql file you loaded earlier. that seemed to calculate stats

File: dbSetup/models/views/batting_stats_view.py (hunk 2 above)
Comment (ickoxii) [reply dbSetup/models/views/batting_stats_view.py:67]: take a look at the create_views_service file in main to get an idea of how i loaded my views.

File: app/routes/grid_routes.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< HEAD
```
Comment (ickoxii) [inline app/routes/grid_routes.py:1]: fix merge conflicts

File: app/static/css/base.css [hunk 5]
```diff
@@ -7,6 +7,7 @@
   --beige: #eaf2d7ff;
   --light-blue: #b3dee2ff;
   --jet: #333333ff;
+<<<<<<< HEAD
```
Comment (ickoxii) [inline app/static/css/base.css:10]: fix merge conflicts

File: app/templates/base.html [hunk 6]
```diff
@@ -37,7 +37,15 @@
             {% if current_user.role == 'ADMIN' %}
                 <a class="navbar-link {% if request.path == '/admin/users' %}navbar-link-active{% endif %}" href="{{ url_for('admin_routes.users') }}">Users</a>
             {% endif %}
+<<<<<<< HEAD
```
Comment (ickoxii) [inline app/templates/base.html:40]: more conflicts

File: app/templates/immaculate_grid.html [hunk 7]
```diff
@@ -3,6 +3,7 @@
 {% from "_macros.html" import prompt %}
 
 {% block content %}
+<<<<<<< HEAD
```
Comment (ickoxii) [inline app/templates/immaculate_grid.html:6]: fix conflicts

File: app/static/css/grid.css [hunk 8]
```diff
@@ -1,157 +1,137 @@
-/* Container for the whole page */
-.solver-container {
-    display: flex;
-    flex-direction: column;
-    align-items: center;
+/* Define CSS HEX colors as variables */
+:root {
```
Comment (brendonnewt) [inline app/static/css/grid.css:2]: I'm assuming Icko did this, but this is the wrong file. The base.css should be in base.css, not grid.css

**Review by brendonnewt — CHANGES_REQUESTED** [review]
I ran some tests using this, and I ran into a few issues. Firstly, I don't think the joins are joining 1-to-1 as there are multiple entries for each player. Secondly, the wOBA, wRCplus, and BsR are giving incorrect output. I'm not sure about the rest, these are just the first I noticed. 

Please try select * from battingstatsview where name = 'Steven Kwan' and yearid = 2022;

Also, if you could structure the playerid, teamID, yearID, stint, etc the same as Icko that would be great for consistency. We don't really need the player name (as multiple players can have the same name) or the team_name (that can be retrieved using the yearid and teamID). Furthermore, try to follow a similar naming scheme to his as well as that matches the rest of the database, and we want these views to be usable without having to go check their field names all the time.

To summarize, there are duplicates, some calculations are incorrect, and the fields need to be tweaked to match the present dbms conventions.

**Comment by brendonnewt** [comment]
> I ran some tests using this, and I ran into a few issues. Firstly, I don't think the joins are joining 1-to-1 as there are multiple entries for each player. Secondly, the wOBA, wRCplus, and BsR are giving incorrect output. I'm not sure about the rest, these are just the first I noticed.
> 
> Please try select * from battingstatsview where name = 'Steven Kwan' and yearid = 2022;
> 
> Also, if you could structure the playerid, teamID, yearID, stint, etc the same as Icko that would be great for consistency. We don't really need the player name (as multiple players can have the same name) or the team_name (that can be retrieved using the yearid and teamID). Furthermore, try to follow a similar naming scheme to his as well as that matches the rest of the database, and we want these views to be usable without having to go check their field names all the time.
> 
> To summarize, there are duplicates, some calculations are incorrect, and the fields need to be tweaked to match the present dbms conventions.

Coming back to this, if they have zero plate appearances (ie. Pitchers), they shouldn't be in this view. Be sure to only include batters in the batting stats please.

**Comment by cateburglar** [comment]
> I ran some tests using this, and I ran into a few issues. Firstly, I don't think the joins are joining 1-to-1 as there are multiple entries for each player. Secondly, the wOBA, wRCplus, and BsR are giving incorrect output. I'm not sure about the rest, these are just the first I noticed.
> 
> Please try select * from battingstatsview where name = 'Steven Kwan' and yearid = 2022;
> 
> Also, if you could structure the playerid, teamID, yearID, stint, etc the same as Icko that would be great for consistency. We don't really need the player name (as multiple players can have the same name) or the team_name (that can be retrieved using the yearid and teamID). Furthermore, try to follow a similar naming scheme to his as well as that matches the rest of the database, and we want these views to be usable without having to go check their field names all the time.
> 
> To summarize, there are duplicates, some calculations are incorrect, and the fields need to be tweaked to match the present dbms conventions.

no please keep name in it!!! the point of the batting stats view was to be able to have all the info we wanted to display in one place , and keeping name in this view makes it easier for the depthchart/team summary

File: dbSetup/services/create_views_service.py [hunk 9]
```diff
@@ -152,3 +152,89 @@ def create_pitching_stats_view():
...
+            b.b_HR AS HR,
+            b.b_SB AS SB,
+            (b.b_BB / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) * 100 AS `BB%`,
+            (b.b_SO / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) * 100 AS `K%`,
+            (b.b_H - b.b_HR) / (b.b_AB - b.b_SO - b.b_HR + b.b_SF) AS BABIP,
+            (b.b_H / b.b_AB) AS AVG,
+            ((b.b_H - (b.b_2B + b.b_3B + b.b_HR)) + (2 * b.b_2B) + (3 * b.b_3B) + (4 * b.b_HR)) / b.b_AB AS SLG,
+            (((b.b_H - (b.b_2B + b.b_3B + b.b_HR)) + (2 * b.b_2B) + (3 * b.b_3B) + (4 * b.b_HR)) / b.b_AB) - (b.b_H / b.b_AB) AS ISO,
+            (b.b_H - (b.b_2B + b.b_3B + b.b_HR)) AS b_1B,
+            (((0.69 * b.b_BB) + (0.72 * b.b_HBP) + (0.88 * (b.b_H - (b.b_2B + b.b_3B + b.b_HR))) + (1.24 * b.b_2B) + (1.56 * b.b_3B) + (2.0 * b.b_HR)) / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) AS wOBA,
```
Comment (ickoxii) [inline dbSetup/services/create_views_service.py:209]: wOBA can be 0.00, not null

File: dbSetup/services/create_views_service.py [hunk 10]
```diff
@@ -152,3 +152,89 @@ def create_pitching_stats_view():
...
+            ((((((0.69 * b.b_BB) + (0.72 * b.b_HBP) + (0.88 * (b.b_H - (b.b_2B + b.b_3B + b.b_HR))) + (1.24 * b.b_2B) + (1.56 * b.b_3B) + (2.0 * b.b_HR)) / (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF)) - 0.320) / 1.25 * (b.b_AB + b.b_BB + b.b_HBP + b.b_SH + b.b_SF) + 6500) / 4000 * 100) AS wRCplus,
```
Comment (ickoxii) [inline dbSetup/services/create_views_service.py:210]: wRCplus, 0 has an actual meaning because it allows negative values. However, it does allow null values

File: dbSetup/services/create_views_service.py [hunk 11]
```diff
@@ -152,3 +152,89 @@ def create_pitching_stats_view():
...
+            ((b.b_SB - b.b_CS) * 0.2) AS BsR,
+            (f.f_PO + f.f_A) AS Total_Defensive_Plays,
+            ((f.f_PO + f.f_A) * 0.1 + f.f_ZR * 0.2 + f.f_DP * 0.5) AS FRAA,
+            b.yearID AS YearID,
+            t.teamID AS TeamID,
+            t.team_name AS Team_Name,
+            f.position AS position,
+            b.stint AS stint
+        FROM
+            batting b
```
Comment (ickoxii) [inline dbSetup/services/create_views_service.py:220]: fuck it natural join everything

Comment (ickoxii) [reply dbSetup/services/create_views_service.py:220]: if that doesn't work join using explicit values (all common attributes)

File: dbSetup/services/create_views_service.py [hunk 12]
```diff
@@ -152,3 +152,89 @@ def create_pitching_stats_view():
...
+        JOIN
+            people p ON b.playerID = p.playerID
+        JOIN
+            appearances a ON b.playerID = a.playerID AND b.yearID = a.yearID
+        JOIN
+            teams t ON b.teamID = t.teamID AND b.yearID = t.yearID
```
Comment (ickoxii) [inline dbSetup/services/create_views_service.py:226]: no need to join teams
